userspace/libscap/engine/bpf/scap_bpf.c

```diff
@@ -699,134 +702,152 @@ static int32_t load_bpf_file(
 		return SCAP_FAILURE;
 	}
 
-	int program_fd = open(handle->m_filepath, O_RDONLY, 0);
-	if(program_fd < 0)
-	{
-		char buf[SCAP_LASTERR_SIZE];
-		snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "can't open BPF probe '%s': %s", handle->m_filepath, scap_strerror_r(buf, errno));
-		return SCAP_FAILURE;
-	}
-
-	Elf *elf = elf_begin(program_fd, ELF_C_READ_MMAP_PRIVATE, NULL);
-	if(!elf)
+	if (!handle->elf)
 	{
-		snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "can't read ELF format");
-		goto cleanup;
-	}
-
-	GElf_Ehdr ehdr;
-	if(gelf_getehdr(elf, &ehdr) != &ehdr)
-	{
-		snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "can't read ELF header");
-		goto cleanup;
-	}
-
-	for(j = 0; j < ehdr.e_shnum; ++j)
-	{
-		if(get_elf_section(elf, j, &ehdr, &shname, &shdr, &data) != SCAP_SUCCESS)
+		handle->program_fd = open(handle->m_filepath, O_RDONLY, 0);
```

Comment:
ELF data is now kept alive in handle, so that we can easily set/unset tracepoints without re-parsing all the elf.
Note that reparsing all the elf resetted all shared maps.